concrete/src/Attribute/Key/SearchIndexer/StandardSearchIndexer.php

```diff
@@ -32,6 +32,71 @@ protected function getIndexEntryColumn(Key $key, $subKey = false)
     }
 
     /**
+     * For certain fields (eg TEXT) Doctrine uses the length of the longest column to determine what field type to use.
+     * For search indexing even if we may not currently have something long in a column,
+     * we need the longest possible column so that we don't truncate any data.
+     *
+     * @param array $options
+     *
+     * @return array
+     */
+    private function setTypeLength($options)
+    {
+        // If we have explicitly set a length, use it
+        if ($options['length']) {
+            return $options;
+        }
+        if ($options['type']->getName() == 'text') {
+            $options['length'] = 4294967295; // This forces Doctrine to use `LONGTEXT` instead of `TINYTEXT`
+        }
+
+        return $options;
+    }
+
+    /**
+     * Refresh the Search Index columns (if there are schema changes for example).
+     *
+     * @param CategoryInterface $category
+     * @param AttributeKeyInterface $key
+     */
+    public function refreshSearchIndexKeyColumns(CategoryInterface $category, AttributeKeyInterface $key)
+    {
+        $controller = $key->getController();
+        $definition = $controller->getSearchIndexFieldDefinition();
+        $sm = $this->connection->getSchemaManager();
+        $fromTable = $sm->listTableDetails($category->getIndexedSearchTable());
+        $toTable = $sm->listTableDetails($category->getIndexedSearchTable());
+
+        if (isset($definition['type'])) {
+            $options = [
+                'type' => Type::getType($definition['type']),
+            ];
+            $options = array_merge($options, $definition['options']);
+            $options = $this->setTypeLength($options);
+            $toTable->changeColumn('ak_' . $key->getAttributeKeyHandle(), $options);
+        } else {
+            foreach ($definition as $name => $column) {
+                $options = [
+                    'type' => Type::getType($column['type']),
+                ];
+                $options = array_merge($options, $column['options']);
+                $options = $this->setTypeLength($options);
+                $toTable->changeColumn('ak_' . $key->getAttributeKeyHandle() . '_' . $name, $options);
+            }
+        }
+        $comparator = new \Doctrine\DBAL\Schema\Comparator();
```

Comment:
A good way to illustrate why creating dependencies within methods can be a bad idea is trying to implement proper tests.

If you wanted to test this method you are required to also test Comparator. You can see that by breaking the comparator class, it will also break any tests you write for this method.

In order to avoid that, you want to push the responsibility of fulfilling the dependency to the caller. Either by making it a hard dependency of the class (Adding a constructor argument) or by making it a hard dependency of the method (Adding an argument to the method).

That way if you want to test this method, you just provide a minimal functionality mock of the $comparator that doesn't actually do anything but allow for testing.

---
Ignoring patterns and design, it will also make your migration quite a lot faster since you'll create the comparator only once. With the current code, you create one comparator per call so the migration could make hundreds of thousands of them for a large site.